### ISSUE TITLE
Receiving Invalid data message - all must be length: 22 from useNativeDriver scroll event

### ISSUE THREAD

**Issue by ryanscottaudio** [facebook/react-native#11760]
### Description

I'm attempting to upgrade from v0.35.0 to v0.39.0 so that I can take advantage of the `useNativeDriver` property on a ton of animations. I'm running into an issue when I attempt to use it on a scroll event in an AnimatedListView (the result of `Animated.createAnimatedComponent(ListView)`). It works correctly for the most part, but if I try to make any action besides scrolling I get hit with the following error message in Xcode: `Invalid data message - all must be length: x`; `x` changes every time and usually stays under about 30. I noticed that Xcode is also complaining about trying to `JSON.stringify` a `cyclic structure`, but I haven't changed anything else in my code, so I'm not sure how that could be happening.

This also happens from just scrolling sometimes.

Example of full error message:
```
2017-01-06 16:38:21.723 [error][tid:com.facebook.react.JavaScript][RCTBatchedBridge.m:958] Invalid data message - all must be length: 22
2017-01-06 16:38:21.729 [error][tid:com.facebook.react.JavaScript][RCTBatchedBridge.m:958] Invalid data message - all must be length: 22
2017-01-06 16:38:21.742 [error][tid:com.facebook.react.JavaScript] JSON.stringify cannot serialize cyclic structures.
2017-01-06 16:38:21.750 [fatal][tid:com.facebook.react.ExceptionsManagerQueue] Unhandled JS Exception: JSON.stringify cannot serialize cyclic structures.
2017-01-06 16:38:22.040 [error][tid:com.facebook.react.JavaScript] JSON.stringify cannot serialize cyclic structures.
2017-01-06 16:38:22.045 [error][tid:com.facebook.react.JavaScript][RCTBatchedBridge.m:958] Invalid data message - all must be length: 22
2017-01-06 16:38:22.049 [error][tid:com.facebook.react.JavaScript][RCTBatchedBridge.m:958] Invalid data message - all must be length: 22
```

### Reproduction
index.js
```
import React, { Component, PropTypes } from 'react';
import {
  AppState,
  Animated,
  LayoutAnimation,
  ListView,
  StyleSheet,
  View
} from 'react-native';

import TopicsHeader from './TopicsHeader';
import TopicCard from './TopicCard';
import Modals from '../Modals';
import { bottomBarHeight } from '../Header';

import { track } from '../../tracking';
import { topics } from '../../config';

const styles = StyleSheet.create({
  container: {
    flex: 1
  }
});

const AnimatedListView = Animated.createAnimatedComponent(ListView);

class TopicsView extends Component {
  constructor(props) {
    super();

    const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });

    this.scrollY = new Animated.Value(0);
    this.renderRow = this.renderRow.bind(this);
    this.state = {
      foreground: props.foreground,
      dataSource: ds.cloneWithRows(topics)
    };
  }

  componentWillMount() {
    LayoutAnimation.easeInEaseOut();
  }

  componentDidMount() {
    track('Topic Feed - Open');
    AppState.addEventListener('change', (appState) => {
      if (appState === 'active') {
        this.setState({ foreground: true });
      } else {
        this.setState({ foreground: false });
      }
    });
  }

  renderRow({ name, tag, description, headlineImage,
    backgroundVideo, backgroundImage, firstFrame, subscribed }, sectionId, rowIndex) {
    const idx = Number(rowIndex);
    const { foreground } = this.state;
    const shouldLoadVideo = foreground;

    return (
      <TopicCard
        name={name}
        tag={tag}
        description={description}
        headlineImage={headlineImage}
        backgroundVideo={backgroundVideo}
        backgroundImage={backgroundImage}
        firstFrame={firstFrame}
        subscribed={this.props.topics[tag]}
        toNotificationsFeed={() => {
          track('Notification Feed - Open', { 'Topic Tag': topics[rowIndex].tag });
          return this.props.dispatch(this.props.actions.navigatorForward({
            key: `NotificationsFeedView_${rowIndex}`,
            passProps: {
              idx: rowIndex
            }
          }));
        }}
        toggleSubscription={() => this.props.dispatch(
          this.props.actions.attemptToToggleSubscription([rowIndex])
        )}
        rowIndex={idx}
        totalRows={topics.length}
        scrollY={this.scrollY}
        shouldLoadVideo={shouldLoadVideo}
      />
    );
  }

  render() {
    return (
      <View style={styles.container}>
        <Modals {...this.props} />
        <TopicsHeader
          {...this.props}
          scrollY={this.scrollY}
        />
        <AnimatedListView
          dataSource={this.state.dataSource}
          renderRow={this.renderRow}
          style={{ flex: 1, backgroundColor: 'black' }}
          onScroll={
            Animated.event([{
              nativeEvent: { contentOffset: { y: this.scrollY } }
            }], {
              useNativeDriver: true
            })
          }
          scrollEventThrottle={1}
          showsVerticalScrollIndicator={false}
          renderHeader={() => <View style={{ height: bottomBarHeight }} />}
          initialListSize={topics.length}
          pageSize={1}
        />
      </View>
    );
  }
}

TopicsView.propTypes = {
  dispatch: PropTypes.func.isRequired
};

export default TopicsView;
```

TopicCard.js
```
import React, { Component, PropTypes } from 'react';
import {
  Animated,
  Dimensions,
  Image,
  StyleSheet,
  Text,
  TouchableWithoutFeedback,
  TouchableOpacity,
  View
} from 'react-native';
import Video from 'react-native-video';
import LinearGradient from 'react-native-linear-gradient';
import { bottomBarHeight, topBarHeight } from '../Header';
import { colors, fontFamily } from '../../utils/styles';
// import { isOldDevice } from '../../utils/helpers';
const isOldDevice = true;

const AnimatedVideo = Animated.createAnimatedComponent(Video);
const AnimatedLinearGradient = Animated.createAnimatedComponent(LinearGradient);
const { width, height } = Dimensions.get('window');
export const cardHeight = width * (467.2 / 375);
const headerHeight = bottomBarHeight + topBarHeight;

const parallaxCoefficient = 1.1;
export const videoHeight = cardHeight * parallaxCoefficient;
const marginBetween = 15;

const styles = StyleSheet.create({
  topicCard: {
    height: cardHeight,
    backgroundColor: 'black',
    overflow: 'hidden'
  },
  topicCardBackground: {
    position: 'absolute',
    height: videoHeight,
    width,
    top: 0,
    left: 0,
    right: 0
  },
  topicCardDetails: {
    width,
    position: 'absolute',
    bottom: 0,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  },
  topicCardHeadlineImage: {
    margin: marginBetween
  },
  subscribeButton: {
    flexDirection: 'row',
    alignItems: 'center',
    borderColor: colors.blue,
    borderWidth: 2,
    padding: 10,
    justifyContent: 'center',
    marginTop: marginBetween,
    marginBottom: marginBetween,
    height: 55,
    width: width * 0.65,
    backgroundColor: 'rgba(0,0,0,0.25)'
  },
  subscribeButtonActive: {
    backgroundColor: colors.blue
  },
  checkIcon: {
    marginLeft: 5,
    height: 11,
    width: 11
  },
  plusIcon: {
    marginLeft: 5,
    height: 11,
    width: 11
  },
  subscribeButtonText: {
    fontFamily: fontFamily.proximaNova.semibold,
    fontSize: 16,
    letterSpacing: 1,
    backgroundColor: 'transparent'
  },
  subscribeButtonTextInactive: {
    color: colors.blue
  },
  subscribeButtonTextActive: {
    color: 'white'
  },
  feedButton: {
    marginTop: marginBetween,
    marginBottom: marginBetween * 2,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  feedButtonText: {
    color: 'white',
    fontFamily: fontFamily.proximaNova.semibold,
    fontSize: 16,
    backgroundColor: 'transparent'
  },
  feedButtonIcon: {
    marginLeft: 6,
    height: 18,
    width: 18
  }
});

class TopicCard extends Component {
  constructor(props) {
    super();

    this.state = {
      animations: { videoOpacity: new Animated.Value(0) }
    };
    this.videoEl = null;
    this.fadeInVideo = this.fadeInVideo.bind(this);
    this.restartVideo = this.restartVideo.bind(this);
  }

  fadeInVideo(e) {
    Animated.timing(
      this.state.animations.videoOpacity,
      {
        toValue: 1,
        duration: 400,
        useNativeDriver: true
      }
    ).start();
  }

  restartVideo() {
    this.videoEl._component.seek(0);
  }

  componentDidUpdate(nextProps) {
    if (nextProps.shouldLoadVideo !== this.props.shouldLoadVideo) {
      Animated.timing(
        this.state.animations.videoOpacity,
        {
          toValue: 0,
          duration: 0,
          useNativeDriver: true
        }
      ).start();
    }
  }

  render() {
    const {
      subscribed, backgroundVideo, backgroundImage, firstFrame, headlineImage,
      toggleSubscription, toNotificationsFeed, rowIndex, totalRows,
      scrollY, shouldLoadVideo, name
    } = this.props;

    const offset = rowIndex * cardHeight;

    const minScrollY = 0;
    const maxScrollY = totalRows * cardHeight + headerHeight - height;

    const containerStyle = [
      styles.topicCard,
      (rowIndex === 0 || rowIndex === totalRows - 1) && {
        zIndex: -1,
        overflow: 'visible'
      }
    ];

    const defaultParallaxRange = {
      inputRange: [offset, offset + (height - headerHeight - cardHeight)],
      outputRange: [0, videoHeight - cardHeight]
    };
    const firstParallaxRange = {
      inputRange: [minScrollY].concat(defaultParallaxRange.inputRange),
      outputRange: [0].concat(defaultParallaxRange.outputRange)
    };
    const lastParallaxRange = {
      inputRange: [].concat(defaultParallaxRange.inputRange).concat([maxScrollY + height - headerHeight]),
      outputRange: [].concat(defaultParallaxRange.outputRange).concat([videoHeight - cardHeight])
    };

    const parallaxRange = rowIndex === 0 ? firstParallaxRange
      : rowIndex === totalRows - 1 ? lastParallaxRange
      : defaultParallaxRange;

    const backgroundStyle = [
      styles.topicCardBackground,
      {
        transform: [
          { translateY: scrollY.interpolate({
            inputRange: [offset, offset + (height - headerHeight - cardHeight)],
            outputRange: [0, videoHeight - cardHeight]
          }) },
          { scale: rowIndex === 0 ? scrollY.interpolate({
            inputRange: [minScrollY - cardHeight + headerHeight, minScrollY, maxScrollY],
            outputRange: [2, 1, 1]
          }) : rowIndex === totalRows - 1 ? scrollY.interpolate({
            inputRange: [minScrollY, maxScrollY, maxScrollY + cardHeight - headerHeight],
            outputRange: [1, 1, 2]
          }) : 1 }
        ]
      }
    ];

    const detailsStyle = [
      styles.topicCardDetails,
      rowIndex === totalRows - 1 && { transform: [{ translateY: scrollY.interpolate({
        inputRange: [maxScrollY - 1, maxScrollY, maxScrollY + 1],
        outputRange: [1, 1, 2]
      }) }] }
    ];

    const subscribeButtonStyle = [
      styles.subscribeButton,
      subscribed ? styles.subscribeButtonActive : null
    ];
    const subscribeButtonTextStyle = [
      styles.subscribeButtonText,
      subscribed ? styles.subscribeButtonTextActive : styles.subscribeButtonTextInactive
    ];

    const subscribeButtonIcon = subscribed
      ? <Image resizeMode='contain' style={styles.checkIcon} source={require('../../../assets/icons/check.png')} />
    : <Image resizeMode='contain' style={styles.plusIcon} source={require('../../../assets/icons/plus.png')} />;
    const text = subscribed ? 'following' : 'follow';

    return (
      <Animated.View style={containerStyle}>
        <TouchableWithoutFeedback
          accessibilityLabel={`go to ${name} notifications feed`}
          accessibilityTraits={'link'}
          onPress={() => toNotificationsFeed(rowIndex)}
        >
          <Animated.Image
            source={isOldDevice ? backgroundImage : firstFrame}
            style={backgroundStyle}
            resizeMode='cover'
          />
        </TouchableWithoutFeedback>
        {!isOldDevice && shouldLoadVideo && <AnimatedVideo
          style={[
            backgroundStyle,
            { opacity: this.state.animations.videoOpacity }
          ]}
          source={backgroundVideo}
          rate={1}
          resizeMode='cover'
          onLoad={this.fadeInVideo}
          repeat={true}
          muted={true}
          playWhenInactive={true}
          pointerEvents='none'
          ref={node => (this.videoEl = node)}
        />}
        <AnimatedLinearGradient
          colors={['transparent', 'rgba(0, 0, 0, 1)']}
          style={detailsStyle}
          pointerEvents='box-none'
        >
          <Image
            style={styles.topicCardHeadlineImage}
            source={headlineImage}
            pointerEvents='none'
          />

          <TouchableOpacity
            accessibilityLabel={`${subscribed ? 'un' : ''}follow ${name} notifications`}
            accessibilityTraits={'button'}
            onPress={() => toggleSubscription([rowIndex])}
            underlayColor='transparent'
          >
            <View style={subscribeButtonStyle}>
              <Text
                style={subscribeButtonTextStyle}
                allowFontScaling={false}
              >
                {text.toUpperCase()}
              </Text>
              {subscribeButtonIcon}
            </View>
          </TouchableOpacity>

          <TouchableOpacity
            accessibilityLabel={`go to ${name} notifications feed`}
            accessibilityTraits={'link'}
            onPress={() => toNotificationsFeed(rowIndex)}
            underlayColor='transparent'
          >
            <View style={styles.feedButton}>
              <Text
                style={styles.feedButtonText}
                allowFontScaling={false}
              >
                LATEST NOTIFICATIONS
              </Text>
              <Image
                style={styles.feedButtonIcon}
                source={require('../../../assets/icons/arrow-right-white.png')}
                resizeMode='contain'
              />
            </View>
          </TouchableOpacity>
        </AnimatedLinearGradient>
      </Animated.View>
    );
  }
}

TopicCard.PropTypes = {
  name: PropTypes.string.isRequired,
  tag: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  subscribed: PropTypes.bool.isRequired,
  headlineImage: PropTypes.object.isRequired,
  backgroundVideo: PropTypes.object.isRequired,
  toggleSubscription: PropTypes.func.isRequired,
  rowIndex: PropTypes.number.isRequired,
  toNotificationsFeed: PropTypes.func.isRequired,
  topicsLength: PropTypes.number.isRequired
};

export default TopicCard;
```

### Additional Information

* React Native version: 0.39
* Platform: iOS
* Operating System: macOS

**Comment by teodors** [comment]
I can confirm. I get the same result on android. I think `useNativeDriver` is not currently ready for `ListView`. You can only use it on `ScrollView`.

I did try it with `ListView` adding this.
```
renderScrollComponent={
(props) => <Animated.ScrollView {...props} onScroll={ATTACH ANIMATED EVENT HERE}/>
}
```
But it unfortunately overwrites [_onScroll](https://github.com/facebook/react-native/blob/55427566b620385ed6b96e791094b0b0e252f848/Libraries/CustomComponents/ListView/ListView.js#L665) and some important `ListView` features break down.

**Comment by teodors** [comment]
@ryanscottaudio It seems, it has been fixed in next release version `0.41.0-rc.0` or `0.41.0` depending on when you read this. This can be closed.

**Comment by scarlac** [comment]
Had same issue. Upgraded. Fixed in 0.41.
Maybe we should also get `Animated.ListView` since we also have `Animated.ScrollView`.

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while.

But don't worry, it will live on with Canny! Check out its new home: https://react-native.canny.io/feature-requests/p/receiving-invalid-data-message---all-must-be-length-22-from-usenativedriver-scro
